### PR TITLE
[Internal] Basic performance testing and monitoring

### PR DESCRIPTION
**Pull request by samhouts**
### Description of Change ###

Setting up basic scenarios that will establish a performance baseline in terms of milliseconds to render. We will run these scenarios in our CI pipeline and output warnings if the performance threshold is exceeded to prevent inadvertently affecting performance as we evolve the product.

We have also extracted the existing internal `Performance` class to the `IPerformanceProvider` interface. Core Xamarin.Forms code will call this provider at critical times (e.g., `Element` is changing). The provider can then choose to snapshot any relevant data at that time.

### Bugs Fixed ###

N/A

### API Changes ###

altered `Xamarin.Forms.Internals.Performance`
added `Xamarin.Forms.Internals.IPerformanceProvider`

### Behavioral Changes ###

None

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.ControlGallery.Android/MainApplication.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+using Android.App;
+using Android.OS;
+using Android.Runtime;
+using Plugin.CurrentActivity;
+
+namespace Xamarin.Forms.ControlGallery.Android
+{
+	//You can specify additional application information in this attribute
+    [Application]
+    public class MainApplication : global::Android.App.Application, global::Android.App.Application.IActivityLifecycleCallbacks
```
Comment (samhouts) [inline Xamarin.Forms.ControlGallery.Android/MainApplication.cs]: Added by `DeviceInfo` plugin. Boilerplate.

File: Xamarin.Forms.ControlGallery.Android/PerformanceTrackerRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+using Xamarin.Forms.Controls;
+using Xamarin.Forms.Platform.Android;
+using static Android.Views.ViewTreeObserver;
+using Android.Views;
+
+[assembly: ExportRenderer(typeof(PerformanceTracker), typeof(PerformanceTrackerRenderer))]
+
+namespace Xamarin.Forms.ControlGallery.Android
+{
+	public class PerformanceTrackerRenderer : ViewRenderer, IOnDrawListener
```
Comment (samhouts) [inline Xamarin.Forms.ControlGallery.Android/PerformanceTrackerRenderer.cs]: This lets me know approximately when the view has finally finished rendering so I can stop the clock.

File: Xamarin.Forms.ControlGallery.iOS/PerformanceTrackerRenderer.cs [hunk 3]
```diff
@@ -0,0 +1,999 @@
...
+[assembly: ExportRenderer(typeof(ProgressBar), typeof(PerformanceTrackingProgressBar))]
+[assembly: ExportRenderer(typeof(SearchBar), typeof(PerformanceTrackingSearchBar))]
+[assembly: ExportRenderer(typeof(Slider), typeof(PerformanceTrackingSlider))]
+[assembly: ExportRenderer(typeof(Stepper), typeof(PerformanceTrackingStepper))]
+[assembly: ExportRenderer(typeof(Switch), typeof(PerformanceTrackingSwitch))]
+[assembly: ExportRenderer(typeof(TableView), typeof(PerformanceTrackingTableView))]
+[assembly: ExportRenderer(typeof(TimePicker), typeof(PerformanceTrackingTimePicker))]
+[assembly: ExportRenderer(typeof(WebView), typeof(PerformanceTrackingWebView))]
+
+namespace Xamarin.Forms.ControlGallery.iOS
```
Comment (samhouts) [inline Xamarin.Forms.ControlGallery.iOS/PerformanceTrackerRenderer.cs]: This is doing what the `PerformanceTrackerRenderer` for Android is doing, but in a more invasive way. I can't do the same trick without having every renderer bubble up to the `PerformanceTrackerRenderer`.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceGallery.cs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+		protected override async void Init()
+		{
+			_BuildInfo = GetBuildNumber();
+
+			_DeviceIdentifier = CrossDeviceInfo.Current.Id;
+			_DeviceIdiom = CrossDeviceInfo.Current.Idiom.ToString();
+			_DeviceModel = CrossDeviceInfo.Current.Model;
+#if __ANDROID__ && TEST_EXPERIMENTAL_RENDERERS
+			_DevicePlatform = "Android Fast Renderers";
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceGallery.cs]: This doesn't work. :(

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceGallery.cs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+			nextButton.IsEnabled = true;
+			nextButton.Text = Next;
+		}
+
+		private static string GetBuildNumber()
+		{
+			try
+			{
+				var assembly = typeof(PerformanceGallery).GetTypeInfo().Assembly;
+				var txt = "Xamarin.Forms.Controls.BuildNumber.txt";
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceGallery.cs]: Embedded resource that I inject with the build number from the build server. This lets me categorize the scenario results on the server side.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceGallery.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+					RunningApp.Screenshot(message);
+					if (!warnings.Contains(message))
+						warnings.Add(message);
+				}
+			}
+
+			string testRunReferenceId = GetText(TestRunRefId);
+
+			if (warnings.Any())
+				Assert.Inconclusive($"Performance threshold exceeded.\r\n{string.Join("\r\n", warnings)}\r\nTestRunReferenceId: {testRunReferenceId}");
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceGallery.cs]: Note that we don't fail the test. Expecting these results to be flaky at times.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceTrackerWatcher.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		}
+
+		public async void WaitForComplete()
+		{
+			await Task.Delay(_timeout);
+
+			if (_sagaComplete)
+				return;
+
+			// triggered lastCall: 12/12/2012 12:12:12:000
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceGallery/PerformanceTrackerWatcher.cs]: These comments offer a sample scenario explaining how I expect this saga to work.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceProvider.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Xamarin.Forms.Internals;
+
+namespace Xamarin.Forms.Controls
+{
+	[Preserve(AllMembers = true)]
+	internal class PerformanceProvider : IPerformanceProvider
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/PerformanceProvider.cs]: This is basically what was in the `Performance` class before.

File: Xamarin.Forms.Platform.Android/AppCompat/Platform.cs [hunk 9]
```diff
@@ -165,7 +165,8 @@ void INavigation.RemovePage(Page page)
 
 		SizeRequest IPlatform.GetNativeSize(VisualElement view, double widthConstraint, double heightConstraint)
 		{
-			Performance.Start();
+			var reference = Guid.NewGuid().ToString();
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/AppCompat/Platform.cs]: Now using a reference ID to distinguish multiple calls to the same method during a render.

File: Xamarin.Forms.Platform.iOS/Cells/CellRenderer.cs [hunk 10]
```diff
@@ -12,6 +12,9 @@ public class CellRenderer : IRegisterable
 
 		public virtual UITableViewCell GetCell(Cell item, UITableViewCell reusableCell, UITableView tv)
 		{
+			var reference = Guid.NewGuid().ToString();
```
Comment (samhouts) [inline Xamarin.Forms.Platform.iOS/Cells/CellRenderer.cs]: iOS hadn't been instrumented before. Please note that it is not fair to compare iOS results to Android results if for no other reason than the instrumentation is vastly different in many cases.

File: Xamarin.Forms.Core/Performance.cs [hunk 11]
```diff
@@ -1,95 +1,34 @@
...
 
-			stats.CallCount++;
-			stats.StartTimes.Push(Stopwatch.GetTimestamp());
-		}
+	[EditorBrowsable(EditorBrowsableState.Never)]
+	public class Performance
+	{
+		public static IPerformanceProvider Provider { get; private set; }
 
-		[Conditional("PERF")]
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Performance.cs]: why are you removing the `Conditional` ?

Comment (samhouts) [reply Xamarin.Forms.Core/Performance.cs]: I have replaced the `Conditional` with a null object pattern. If no provider is found, then the code is no-op. The `Conditional` can't be there if we want to allow our users to build their own performance trackers, which was part of my requirements. Core code is compiled by the time it makes it to users.

**Pull request merged**